thorlcr/activities/hashdistrib/thhashdistribslave.cpp

```diff
@@ -140,21 +140,46 @@ class CDistributorBase : implements IHashDistributor, implements IExceptionHandl
              */
             dedupList.sort(*iCompare, 1);
 
-            OwnedConstThorRow prev;
-            for (unsigned i = c; i>0;)
+            if (!keepBestCompare)
             {
-                OwnedConstThorRow row = dedupList.getClear(--i);
-                if ((NULL != prev.get()) && (0 == iCompare->docompare(prev, row)))
+                OwnedConstThorRow prev;
+                for (unsigned i = c; i>0;)
                 {
-                    /* NB: do not alter 'total' size. It represents the amount originally added to the bucket
-                     * which will be deducted when sent.
-                     */
+                    OwnedConstThorRow row = dedupList.getClear(--i);
+                    if ((NULL != prev.get()) && (0 == iCompare->docompare(prev, row)))
+                    {
+                        /* NB: do not alter 'total' size. It represents the amount originally added to the bucket
+                        * which will be deducted when sent.
+                        */
+                    }
+                    else
+                    {
+                        prev.set(row);
+                        rows.enqueue(row.getClear());
+                    }
                 }
-                else
+            }
+            else
+            {
+                OwnedConstThorRow prev;
+                unsigned i = c;
+                if (i)
+                    prev.set(dedupList.getClear(--i));
+                while (i>0)
                 {
+                    OwnedConstThorRow row = dedupList.getClear(--i);
+
+                    if (0 == iCompare->docompare(prev, row))
+                    {
+                        if (keepBestCompare->docompare(prev,row) <= 0)
+                            continue;
+                    }
+                    else
+                        rows.enqueue(prev.getClear());
                     prev.set(row);
```

Comment:
minor: prev.setown(row.getClear()) is more efficient in this case (row is not used later on)